lang/src/main/kotlin/org/partiql/lang/util/NumberExtensions.kt

```diff
@@ -76,10 +82,11 @@ internal fun Number.isZero() = when (this) {
     else -> throw IllegalStateException()
 }
 
+@Suppress("UNCHECKED_CAST")
 /** Provides a narrowing or widening operator on supported numbers. */
-fun Number.coerce(type: Class<out Number>): Number {
+fun <T> Number.coerce(type: Class<T>): T where T : Number {
     val conv = CONVERTERS[type] ?: throw IllegalArgumentException("No converter for $type")
-    return conv(this)
+    return conv(this) as T
 }
```

Comment:
The change here optimizes this helper method, so we don't need to cast manually to a specific subtype of `Number` after calling this method.

Reply:
The compiler recognize the class of the argument `type` and infer the return type of this method as it.

Reply:
`out Number` refers to a class which extends `Number`

Reply:
Now I changed it to generic type `T`, and the `where T : Number` clause specifies that `T` extends `Number`. Since now we have the reference `T` to the type, we can specify the return type of `Number.coerce()` to be `T`